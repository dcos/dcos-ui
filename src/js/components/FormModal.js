import classNames from "classnames/dedupe";
import { Form, Modal } from "reactjs-components";
import React from "react";

const METHODS_TO_BIND = [
  "getTriggerSubmit",
  "handleTriggerSubmit",
  "handleError"
];

class FormModal extends React.Component {
  constructor() {
    super();
    this.triggerSubmit = function() {};

    METHODS_TO_BIND.forEach(method => {
      this[method] = this[method].bind(this);
    });
  }

  componentDidMount() {
    this.focusOnField();
  }

  componentDidUpdate(prevProps) {
    if (prevProps.open === false && this.props.open === true) {
      this.focusOnField();
    }
  }

  handleTriggerSubmit() {
    this.triggerSubmit();
  }

  getTriggerSubmit(trigger) {
    this.triggerSubmit = trigger;
    this.forceUpdate();
  }

  handleNewButtonClick() {
    this.triggerSubmit();
  }

  handleError() {
    this.forceUpdate();
  }

  focusOnField() {
    // Gotta account for animation
    setTimeout(() => {
      const el = this.refs["form-wrapper"];
      if (el) {
        const input = el.querySelector("form input");
        if (input) {
          input.focus();
        }
      }
    }, 100);
  }

  getButtons() {
    return this.props.buttonDefinition.map((buttonDefinition, i) => {
      let buttonClassSet = {
        disabled: this.props.disabled
      };
      buttonClassSet[buttonDefinition.className] = true;
      buttonClassSet = classNames(buttonClassSet);

      let handleOnClick = function() {};
      if (buttonDefinition.isClose) {
        handleOnClick = this.props.onClose;
      }

      if (buttonDefinition.isSubmit) {
        handleOnClick = this.handleTriggerSubmit;
      }

      if (buttonDefinition.onClick) {
        handleOnClick = buttonDefinition.onClick;
      }

      return (
        <button
          className={buttonClassSet}
          disabled={this.props.disabled}
          key={i}
          onClick={handleOnClick}
        >
          {buttonDefinition.text}
        </button>
      );
    });
  }

  getFooter() {
    return (
      <div className="flush-bottom flex flex-direction-top-to-bottom flex-align-items-stretch-screen-small flex-direction-left-to-right-screen-small flex-justify-items-space-between-screen-small">
        {this.getButtons()}
        {this.props.extraFooterContent}
      </div>
    );
  }

  getClassName(hasContentFooter) {
    if (hasContentFooter) {
      return "form";
    }

    return "form flush-bottom";
  }

  getContent() {
    return (
      <div ref="form-wrapper" className={classNames(this.props.contentClasses)}>
        {this.props.children}
        <Form
          className={this.getClassName(!!this.props.contentFooter)}
          definition={this.props.definition}
          triggerSubmit={this.getTriggerSubmit}
          onSubmit={this.props.onSubmit}
          onChange={this.props.onChange}
          onError={this.handleError}
        />
        {this.props.contentFooter}
      </div>
    );
  }

  render() {
    return (
      <Modal
        closeByBackdropClick={!this.props.disabled}
        modalClass="modal modal-small"
        onClose={this.props.onClose}
        open={this.props.open}
        showHeader={false}
        showFooter={true}
        footer={this.getFooter()}
        {...this.props.modalProps}
      >
        {this.getContent()}
      </Modal>
    );
  }
}

FormModal.defaultProps = {
  buttonDefinition: [
    {
      text: "Cancel",
      className: "button button-primary-link flush-left",
      isClose: true
    },
    {
      text: "Create",
      className: "button button-primary",
      isSubmit: true
    }
  ],
  disabled: false,
  extraFooterContent: null,
  onChange() {},
  onClose() {},
  open: false,
  modalProps: {}
};

FormModal.propTypes = {
  buttonDefinition: React.PropTypes.array,
  children: React.PropTypes.node,
  contentClasses: React.PropTypes.oneOfType([
    React.PropTypes.array,
    React.PropTypes.object,
    React.PropTypes.string
  ]),
  contentFooter: React.PropTypes.node,
  definition: React.PropTypes.array,
  disabled: React.PropTypes.bool,
  extraFooterContent: React.PropTypes.node,
  modalProps: React.PropTypes.object,
  onChange: React.PropTypes.func,
  onClose: React.PropTypes.func.isRequired,
  onSubmit: React.PropTypes.func,
  open: React.PropTypes.bool
};

module.exports = FormModal;
