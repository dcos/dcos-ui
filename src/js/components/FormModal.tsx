import { Trans } from "@lingui/macro";
import { i18nMark } from "@lingui/react";
import classNames from "classnames/dedupe";
import { Form, Modal } from "reactjs-components";
import PropTypes from "prop-types";
import * as React from "react";

export default class FormModal extends React.Component {
  static defaultProps = {
    buttonDefinition: [
      {
        text: i18nMark("Cancel"),
        className: "button button-primary-link flush-left",
        isClose: true,
      },
      {
        text: i18nMark("Create"),
        className: "button button-primary",
        isSubmit: true,
      },
    ],
    disabled: false,
    extraFooterContent: null,
    onChange() {},
    onClose() {},
    open: false,
    modalProps: {},
  };
  static propTypes = {
    buttonDefinition: PropTypes.array,
    children: PropTypes.node,
    contentClasses: PropTypes.oneOfType([
      PropTypes.array,
      PropTypes.object,
      PropTypes.string,
    ]),
    contentFooter: PropTypes.node,
    definition: PropTypes.array,
    disabled: PropTypes.bool,
    extraFooterContent: PropTypes.node,
    modalProps: PropTypes.object,
    onChange: PropTypes.func,
    onClose: PropTypes.func.isRequired,
    onSubmit: PropTypes.func,
    open: PropTypes.bool,
  };

  formWrapperRef = React.createRef();

  triggerSubmit = () => {};

  componentDidMount() {
    this.focusOnField();
  }

  componentDidUpdate(prevProps) {
    if (prevProps.open === false && this.props.open === true) {
      this.focusOnField();
    }
  }
  handleTriggerSubmit = () => {
    this.triggerSubmit();
  };
  getTriggerSubmit = (trigger) => {
    this.triggerSubmit = trigger;
    this.forceUpdate();
  };

  handleNewButtonClick() {
    this.triggerSubmit();
  }
  handleError = () => {
    this.forceUpdate();
  };

  focusOnField() {
    // Gotta account for animation
    setTimeout(() => {
      if (this.formWrapperRef && this.formWrapperRef.current) {
        const input = this.formWrapperRef.current.querySelector("form input");
        if (input) {
          input.focus();
        }
      }
    }, 100);
  }

  getButtons() {
    return this.props.buttonDefinition.map((buttonDefinition, i) => {
      let buttonClassSet = {
        disabled: this.props.disabled,
      };
      buttonClassSet[buttonDefinition.className] = true;
      buttonClassSet = classNames(buttonClassSet);

      let handleOnClick = () => {};
      if (buttonDefinition.isClose) {
        handleOnClick = this.props.onClose;
      }

      if (buttonDefinition.isSubmit) {
        handleOnClick = this.handleTriggerSubmit;
      }

      if (buttonDefinition.onClick) {
        handleOnClick = buttonDefinition.onClick;
      }
      return (
        <button
          className={buttonClassSet}
          disabled={this.props.disabled}
          key={i}
          onClick={handleOnClick}
        >
          <Trans id={buttonDefinition.text} />
        </button>
      );
    });
  }

  getFooter() {
    return (
      <div className="flush-bottom flex flex-direction-top-to-bottom flex-align-items-stretch-screen-small flex-direction-left-to-right-screen-small flex-justify-items-space-between-screen-small">
        {this.getButtons()}
        {this.props.extraFooterContent}
      </div>
    );
  }

  getClassName(hasContentFooter) {
    if (hasContentFooter) {
      return "form";
    }

    return "form flush-bottom";
  }

  render() {
    return (
      <Modal
        closeByBackdropClick={!this.props.disabled}
        modalClass="modal modal-small"
        onClose={this.props.onClose}
        open={this.props.open}
        showHeader={false}
        showFooter={true}
        footer={this.getFooter()}
        {...this.props.modalProps}
      >
        <div
          ref={this.formWrapperRef}
          className={classNames(this.props.contentClasses)}
        >
          {this.props.children}
          <Form
            className={this.getClassName(!!this.props.contentFooter)}
            definition={this.props.definition}
            triggerSubmit={this.getTriggerSubmit}
            onSubmit={this.props.onSubmit}
            onChange={this.props.onChange}
            onError={this.handleError}
          />
          {this.props.contentFooter}
        </div>
      </Modal>
    );
  }
}
