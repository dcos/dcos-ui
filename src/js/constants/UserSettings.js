module.exports = {
  SAVED_STATE_KEY: 'savedStates'
};
