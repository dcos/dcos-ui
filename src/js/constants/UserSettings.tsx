export const SAVED_STATE_KEY = "savedStates";
