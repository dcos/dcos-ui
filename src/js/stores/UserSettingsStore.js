import EventEmitter from 'events';

import LocalStorageUtil from '../utils/LocalStorageUtil';

const LOCAL_STORAGE_KEY = 'dcosUserSettings';

function getLocalStorageObject() {
  let localStorageObject = LocalStorageUtil.get(LOCAL_STORAGE_KEY);
  try {
    return JSON.parse(localStorageObject);
  } catch (e) {
    return null;
  }
}

class UserSettingsStore extends EventEmitter {
  constructor() {
    super(...arguments);
  }

  getKey(key) {
    let localStorageObject = getLocalStorageObject();
    if (localStorageObject == null) {
      return null;
    }

    return localStorageObject[key];
  }

  setKey(key, value) {
    let localStorageObject = getLocalStorageObject();
    if (localStorageObject == null) {
      localStorageObject = {};
    }

    localStorageObject[key] = value;
    LocalStorageUtil.set(LOCAL_STORAGE_KEY, JSON.stringify(localStorageObject));
  }

  get storeID() {
    return 'userSettings';
  }
}

module.exports = new UserSettingsStore();
