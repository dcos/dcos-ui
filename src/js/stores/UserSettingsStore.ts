import EventEmitter from "events";

import LocalStorageUtil from "../utils/LocalStorageUtil";

const LOCAL_STORAGE_KEY = "dcosUserSettings";
const REFRESH_RATE_SETTING = "RefreshRate";

function getLocalStorageObject() {
  const localStorageObject = LocalStorageUtil.get(LOCAL_STORAGE_KEY);
  if (localStorageObject != null) {
    try {
      return JSON.parse(localStorageObject);
    } catch (e) {
      return null;
    }
  }
  return null;
}

class UserSettingsStore extends EventEmitter {
  getKey(key: string): unknown {
    const localStorageObject = getLocalStorageObject();
    if (localStorageObject == null) {
      return null;
    }

    return localStorageObject[key];
  }

  setKey(key: string, value: unknown): void {
    let localStorageObject = getLocalStorageObject();
    if (localStorageObject == null) {
      localStorageObject = {};
    }

    localStorageObject[key] = value;
    LocalStorageUtil.set(LOCAL_STORAGE_KEY, JSON.stringify(localStorageObject));
  }

  get storeID() {
    return "userSettings";
  }

  get RefreshRateSetting(): number | null {
    const settings = this.getKey(REFRESH_RATE_SETTING);

    return typeof settings === "number" ? settings : null;
  }

  setRefreshRateSetting(refreshRate: number) {
    this.setKey(REFRESH_RATE_SETTING, refreshRate);
  }
}

export default new UserSettingsStore();
