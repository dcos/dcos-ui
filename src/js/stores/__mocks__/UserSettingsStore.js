var UserSettingsStore = jest.genMockFromModule('../UserSettingsStore');

var _data = {};

/**
 * Sets mock data for UserSettingsStore.getKey to return.
 * @param {String} key the retrieval key
 * @param {Object} response the data to return for the given key
 * @memberOf __mocks__/UserSettingsStore
 */
function __setKeyResponse(key, response) {
  _data[key] = response;
}

/**
 * Clears all mock data from the UserSettingsStore.
 * @memberOf __mocks__/UserSettingsStore
 */
function __reset() {
  _data = {};
}

/**
 * Mocks UserSettingsStore.getKey
 * @param {String} key the retrieval key
 * @returns {Object} mock data if present, otherwise an empty object.
 * @memberOf __mocks__/UserSettingsStore
 */
function getKey(key) {
  if (_data[key] == null) {
    return {};
  }
  return _data[key];
}

UserSettingsStore.__setKeyResponse = __setKeyResponse;
UserSettingsStore.__reset = __reset;
UserSettingsStore.getKey = getKey;

module.exports = UserSettingsStore;
