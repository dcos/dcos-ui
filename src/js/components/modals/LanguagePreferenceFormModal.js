import React from "react";
import { Trans, t } from "@lingui/macro";
import { i18nMark, withI18n } from "@lingui/react";
import Languages from "#SRC/js/constants/Languages";
import LanguageModalActions from "#SRC/js/events/LanguageModalActions";
import FormModal from "../FormModal";
import ModalHeading from "../modals/ModalHeading";
import UserLanguageStore from "../../stores/UserLanguageStore";

export class LanguagePreferenceFormModalComponent extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      isOpen: props.isOpen || false
    };

    this.handleClose = this.handleClose.bind(this);
    this.handleLanguagePrefSubmit = this.handleLanguagePrefSubmit.bind(this);
  }

  componentWillReceiveProps(nextProps) {
    const { isOpen } = nextProps;
    if (isOpen !== this.state.isOpen) {
      this.setState({ isOpen });
    }
  }

  handleClose() {
    LanguageModalActions.close();
    this.setState({ isOpen: false });
  }

  handleLanguagePrefSubmit(formData) {
    UserLanguageStore.set(formData.language);

    this.handleClose();
  }

  getLanguagePreferenceDefinition() {
    const languages = Object.keys(Languages).map(lang => {
      return { html: Languages[lang], id: lang };
    });

    return [
      {
        fieldType: "select",
        label: this.props.i18n
          ? this.props.i18n._(t`Select your language`)
          : "Select your language",
        showLabel: true,
        options: languages,
        value: UserLanguageStore.get(),
        name: "language",
        formElementClass: "languageDropdown-wrapper"
      }
    ];
  }

  getButtonDefinition() {
    return [
      {
        text: i18nMark("Cancel"),
        className: "button button-primary-link flush-left",
        isClose: true
      },
      {
        text: i18nMark("Save"),
        className: "button button-primary",
        isSubmit: true
      }
    ];
  }

  getHeader() {
    return (
      <ModalHeading>
        <Trans render="span">Language Preference</Trans>
      </ModalHeading>
    );
  }

  render() {
    return (
      <FormModal
        buttonDefinition={this.getButtonDefinition()}
        definition={this.getLanguagePreferenceDefinition()}
        modalProps={{
          header: this.getHeader(),
          showHeader: true
        }}
        onClose={this.handleClose}
        onSubmit={this.handleLanguagePrefSubmit}
        open={this.state.isOpen}
      />
    );
  }
}

export default withI18n()(LanguagePreferenceFormModalComponent);
