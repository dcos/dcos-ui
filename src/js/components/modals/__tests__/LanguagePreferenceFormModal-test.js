import React from "react";
import { shallow } from "enzyme";
import { LanguagePreferenceFormModalComponent } from "../LanguagePreferenceFormModal";
import UserSettingsStore from "../../../stores/UserSettingsStore";
import { SAVED_STATE_KEY } from "../../../constants/UserSettings";
import FormModal from "../../FormModal";

const defaultLanguage = "en";
const newLanguage = "zh";
const mockFormData = { language: newLanguage };

describe("LanguagePreferenceFormModalComponent", function() {
  beforeEach(function() {
    UserSettingsStore.setKey(SAVED_STATE_KEY, { language: defaultLanguage });
  });

  it("renders with default language", function() {
    const component = shallow(<LanguagePreferenceFormModalComponent />);
    expect(component).toMatchSnapshot();
  });

  describe("when submitting", function() {
    it("sets isOpen to false", function() {
      const component = shallow(
        <LanguagePreferenceFormModalComponent isOpen={true} />
      );
      expect(component.state("isOpen")).toBe(true);
      component
        .find(FormModal)
        .props()
        .onSubmit(mockFormData);
      expect(component.state("isOpen")).toBe(false);
    });

    it("calls handleLanguagePrefSubmit with new language", function() {
      const handleLanguagePrefSubmitSpy = spyOn(
        LanguagePreferenceFormModalComponent.prototype,
        "handleLanguagePrefSubmit"
      );
      const component = shallow(<LanguagePreferenceFormModalComponent />);
      expect(handleLanguagePrefSubmitSpy).not.toHaveBeenCalled();
      component
        .find(FormModal)
        .props()
        .onSubmit(mockFormData);
      expect(handleLanguagePrefSubmitSpy).toHaveBeenCalledWith(mockFormData);
    });

    it("stores the user's selection in UserSettingsStore", function() {
      const component = shallow(<LanguagePreferenceFormModalComponent />);
      expect(UserSettingsStore.getKey(SAVED_STATE_KEY).language).toBe(
        defaultLanguage
      );
      component
        .find(FormModal)
        .props()
        .onSubmit(mockFormData);
      expect(UserSettingsStore.getKey(SAVED_STATE_KEY).language).toBe(
        newLanguage
      );
    });
  });

  describe("cancel", function() {
    it("sets isOpen to false", function() {
      const component = shallow(
        <LanguagePreferenceFormModalComponent isOpen={true} />
      );
      expect(component.state("isOpen")).toBe(true);
      component
        .find(FormModal)
        .props()
        .onClose();
      expect(component.state("isOpen")).toBe(false);
    });
  });
});
