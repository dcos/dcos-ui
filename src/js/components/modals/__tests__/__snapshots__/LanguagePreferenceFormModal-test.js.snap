// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LanguagePreferenceFormModalComponent renders with default language 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <LanguagePreferenceFormModalComponent />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "buttonDefinition": Array [
        Object {
          "className": "button button-primary-link flush-left",
          "isClose": true,
          "text": "Cancel",
        },
        Object {
          "className": "button button-primary",
          "isSubmit": true,
          "text": "Save",
        },
      ],
      "definition": Array [
        Object {
          "fieldType": "select",
          "formElementClass": "languageDropdown-wrapper",
          "label": "Select your language",
          "name": "language",
          "options": Array [
            Object {
              "html": "English",
              "id": "en",
            },
            Object {
              "html": "中文",
              "id": "zh",
            },
          ],
          "showLabel": true,
          "value": "en",
        },
      ],
      "disabled": false,
      "extraFooterContent": null,
      "modalProps": Object {
        "header": <ModalHeading
          flush={true}
          level={2}
        >
          <WithI18n
            id="Language Preference"
            render="span"
          />
        </ModalHeading>,
        "showHeader": true,
      },
      "onChange": [Function],
      "onClose": [Function],
      "onSubmit": [Function],
      "open": false,
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "buttonDefinition": Array [
          Object {
            "className": "button button-primary-link flush-left",
            "isClose": true,
            "text": "Cancel",
          },
          Object {
            "className": "button button-primary",
            "isSubmit": true,
            "text": "Save",
          },
        ],
        "definition": Array [
          Object {
            "fieldType": "select",
            "formElementClass": "languageDropdown-wrapper",
            "label": "Select your language",
            "name": "language",
            "options": Array [
              Object {
                "html": "English",
                "id": "en",
              },
              Object {
                "html": "中文",
                "id": "zh",
              },
            ],
            "showLabel": true,
            "value": "en",
          },
        ],
        "disabled": false,
        "extraFooterContent": null,
        "modalProps": Object {
          "header": <ModalHeading
            flush={true}
            level={2}
          >
            <WithI18n
              id="Language Preference"
              render="span"
            />
          </ModalHeading>,
          "showHeader": true,
        },
        "onChange": [Function],
        "onClose": [Function],
        "onSubmit": [Function],
        "open": false,
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
